gcc/testsuite/rust.test/xfail_compile/implicit_returns_err2.rs

```diff
@@ -1,9 +1,6 @@
-// { dg-error "expected .* got .*" "" { target { *-*-* } } 0 }
-
 fn test(x: i32) -> i32 {
-    return 1;
-    // { dg-warning "unreachable expression" "" { target *-*-* } .+1 }
-    true
+    return true; // { dg-error "expected .i32. got .bool." }
```

Comment:
I would keep the previous test and add this modification as a new test case. This tests a different thing. Before it checked that `true` is unreachable (warning, but compile-pass) AND has the wrong type (hard error, compile fail), where the first warning should not prevent detecting the second error.

With this change it removes the dependency of those errors.

Before: "This test fails even though the expression is unreachable"
Now: "This test fails because it returns the wrong type and emits a warning"

Reply:
I agree with you. But I will justify my changes. Before this PR, all unreachable statements are removed but unreachable trailing expressions are kept. So here is a type-mistached error. But it is not reasonable at all since the unreachable expression may refer removed variables:
```rust
fn foo() -> i32 {
    return 1;
    let tmp: f32 = 0.0; // removed because it is unreachable
    tmp // gccrs cannot find tmp since it is removed
}
```
So in this PR, unreachable statements AND trailing expressions are removed. Thus the type-mismatched error disappears and the test is modified accordingly.

Reply:
I'm not saying that your changes are wrong. I'm just saying that you should keep the old test and that it should still fail the compilation. It doesn't _really_ matter what/how many errors/warnings are emitted as long as it still fails.

Generally removing regression tests isn't something you should do. This practically removes one test and replaces it with another. I would just add the new test and keep the old test (with the `dg` comments adjusted).

Reply:
OK, I understand now. But maybe the key problem is that the original test won't fail anymore, because the trailing expression is removed.

In my opinion, the type check against unreachable statements and expressions is totally broken and unreasonable before this PR. It skips all unreachable statements and checks the unreachable expression. The original test is just an accident so we have got the right error message. (You can see the bad case in the example above.)

As far as I'm concerned, I would rather make the type check skip all unreachable statements and the trailing expressions temporarily (and then the original test does compile). How do you think about it?

Reply:
> the type check against unreachable statements and expressions is totally broken and unreasonable before this PR

AKAICT it complies with the Rust reference and `rustc`. You can't skip unreachable code in the semantic analysis as long as the `unreachable` analysis only produces a warning and not a hard error.

> As far as I'm concerned, I would rather make the type check skip all unreachable statements AND the trailing expressions temporarily (and then the original test does compile). How do you think about it?

This is against the Rust reference though, so it breaks the compiler. This test _must not_ compile.

Reply:
Indeed, that would be really awkward as some unreachable code would build fine until one day it's not unreachable anymore and causes type errors without having touched it...

Reply:
The original implementation is also against the Rust reference. After all it skips all unreachable statements. So maybe I should 
either do not change it at all, or make it completely right?

Reply:
> Indeed, that would be really awkward as some unreachable code would build fine until one day it's not unreachable anymore and causes type errors without having touched it...

Ah yes, the python way 🐍 

> So maybe I should either do not change it at all, or make it completely right?

No, but you shouldn't break existing regression tests, aka introducing new bugs (that were fixed before hence the regression test).

Reply:
OK. I'll revert the change.

Reply:
This issue should be fixed completely. It is actually very easy to fix (sorry for this since I didn't think it carefully before and I thought it should be a separate issue).